packages/web/src/store/tipping/utils.ts

```diff
@@ -0,0 +1,187 @@
+    }
+    window.localStorage.setItem(RECENT_TIPS_KEY, JSON.stringify(newStorage))
+    return oldestValidTip
+  }
+
+  /**
+   * If recent tip is previously dismissed, then
+   * first try to find user's own tip that is
+   * more recent than dismissed tip.
+   * If not found, then try to find tip more recent
+   * than dismissed tip.
+   * If not found, then try to find tip as recent as that
+   * which was dismissed if the last dismissal is too old.
+   * Otherwise, there is nothing to return.
+   *
+   * Also update local storage if some tip is returned.
+   */
+  if (storage.dismissed) {
+    const ownTip = sortedTips.find(
+      tip => tip.sender_id === userId && tip.slot > storage.minSlot
+    )
+    if (ownTip) {
+      const newStorage = {
+        minSlot: ownTip.slot,
+        dismissed: false,
+        lastDismissalTimestamp: null
+      }
+      window.localStorage.setItem(RECENT_TIPS_KEY, JSON.stringify(newStorage))
+      return ownTip
+    }
+
+    const oldestValidTip = sortedTips.find(tip => tip.slot > storage.minSlot)
+    if (oldestValidTip) {
+      const newStorage = {
+        minSlot: oldestValidTip.slot,
+        dismissed: false,
+        lastDismissalTimestamp: null
+      }
+      window.localStorage.setItem(RECENT_TIPS_KEY, JSON.stringify(newStorage))
+      return oldestValidTip
+    }
+
+    if (
+      storage.lastDismissalTimestamp &&
+      Date.now() - storage.lastDismissalTimestamp >
+        FEED_TIP_DISMISSAL_TIME_LIMIT
+    ) {
+      const oldestValidTip = sortedTips.find(
+        tip => tip.slot === storage.minSlot
+      )
+      if (oldestValidTip) {
+        const newStorage = {
+          minSlot: oldestValidTip.slot,
+          dismissed: false,
+          lastDismissalTimestamp: null
+        }
+        window.localStorage.setItem(RECENT_TIPS_KEY, JSON.stringify(newStorage))
+        return oldestValidTip
+      }
+    }
+
+    return null
+  }
+
+  /**
+   * If recent tip is not dismissed, then
+   * try to find user's own tip as recent or more recent
+   * than previously displayed tip.
+   * If not found, then try to find tip more recent
+   * than previously displayed tip.
+   * Otherwise, there is nothing to return.
+   *
+   * Also update local storage if some tip is returned.
+   */
+  const ownTip = sortedTips.find(
+    tip => tip.sender_id === userId && tip.slot >= storage.minSlot
+  )
+  if (ownTip) {
+    const newStorage = {
+      minSlot: ownTip.slot,
+      dismissed: false,
+      lastDismissalTimestamp: null
+    }
+    window.localStorage.setItem(RECENT_TIPS_KEY, JSON.stringify(newStorage))
+    return ownTip
+  }
+
+  const oldestValidTip = sortedTips.find(tip => tip.slot >= storage.minSlot)
+  if (oldestValidTip) {
+    const newStorage = {
+      minSlot: oldestValidTip.slot,
+      dismissed: false,
+      lastDismissalTimestamp: null
+    }
+    window.localStorage.setItem(RECENT_TIPS_KEY, JSON.stringify(newStorage))
+    return oldestValidTip
+  }
+
+  return null
```

Comment:
Hmm, this is the only part of the logic that I don't think is quite correct – pretty sure if haven't dismissed the tile but there are no newer tips than we saw last time, we should just show the newest thing (rather than showing nothing)?

Reply:
this should not even really be reachable. the check is for `>=` so if a tile had already existed before, its tip should always exists even if there is no new tip after it, which means we won't reach this scenario